File: Основы программирования/Лабораторные работы/18/Pascal/source.pas
program lab_18;

{$MODE TP}
uses U_SORT;

const
	marks_number = 3;

type
	student = record
		name: string[50];
		marks: array [1..marks_number] of byte;
		avg: real;
	end;

	pstudent = ^student;
	students = array[1..65520 div sizeof(student)] of student;
	pstudents = ^students;

{$F+}
function lexicographical(var a, b): boolean; far;

	begin

		lexicographical := pstudent(@a)^.name < pstudent(@b)^.name;

	end;

function nonincreasing(var a, b): boolean; far;

	begin

		nonincreasing := pstudent(@a)^.avg <= pstudent(@b)^.avg;

	end;
{$F-}

function readFile(var arr: pstudents): boolean;

	var
		i, j, sum, counter: byte;
		input: text;
		c: char;

	begin

		readFile := FALSE;

		assign(input, 'input.txt');
		reset(input);

		ReturnNilIfGrowHeapFails := TRUE;

		getmem(arr, sizeof(student) * n);
		i := 1;

		if arr <> nil then begin
			read(input, n);
			while (i <= n) do
				begin
					arr^[i].name := '';
					counter := 0;

					while counter < 3 do
						begin
							read(input, c);
							arr^[i].name := arr^[i].name + c;
							if c = ' ' then
								inc(counter);
						end;

					j := 1;
					sum := 0;

					while j <= marks_number do
						begin
							read(input, arr^[i].marks[j]);
							sum := sum + arr^[i].marks[j];
							inc(j);
						end;

					arr^[i].avg := sum / marks_number;
					readln(input);
					inc(i);
				end;

			readFile := TRUE;
		end;

		close(input);

	end;

procedure writeFile(arr: pstudents);

	var
		i, j: byte;
		output: text;

	begin

		assign(output, 'output.txt');
		rewrite(output);

		for i := 1 to n do
			begin
				write(output, arr^[i].name);
				for j := 1 to marks_number do
					write(output, arr^[i].marks[j], ' ');
				writeln(output, arr^[i].avg:3);
			end;

		close(output);

	end;

var
	arr: pstudents;
	mode: char;

begin

	if (readFile(arr)) then
		begin
			writeln('[L]exicographical or [N]onincreasing?');
			readln(mode);
			case (mode) of
				'L': sort(arr, n, sizeof(student), lexicographical);
				'N': sort(arr, n, sizeof(student), nonincreasing);
				else
					writeln('Error! Input of the initial data!');
			end;

			writeFile(arr);
		end
	else
		writeln('Not enough memory!');

	readln;

end.

File: Основы программирования/Лабораторные работы/18/C/source.c
#include <stdio.h>
#include <stdlib.h>
#include "u_sort.h"

#define marks_number 3

typedef struct {
	char name[50];
	int marks[marks_number];
	float avg;
} student;
	
int lexicographical(void *a, void *b) {
	int i = 0;
	
	for (; (*(student *)a).name[i] && (*(student *)a).name[i] == (*(student *)b).name[i]; i++) ;
	
	return ((*(student *)a).name[i] < (*(student *)b).name[i]);
} 

int nonincreasing(void *a, void *b) {
	return ((*(student *)a).avg > (*(student *)b).avg);
}

void funcStrcat(char *to, char *from) {

	while(*to) to++;
		
	while(*to++ = *from++); 
	
	*(--to) = ' ';
	*(++to) = '\0';

}

size_t readFile(student *arr, unsigned *n) {
	size_t i = 0, j, sum;
	FILE *file;
	char tmp[20];

	if (file = fopen("input.txt", "r")) {
		fscanf(file, "%u\n", &*n);
		if (arr = (student *) malloc (sizeof(student) * *n)) {
			for (; i < *n; i++)  {

				sum = 0;
				arr[i].name[0] = '\0';
				for(j = 0; j < 3; j++) {
					fscanf(file, "%s", tmp); 
					funcStrcat(arr[i].name, tmp);
				}
				for (j = 0; j < marks_number; j++)
				{
					fscanf(file,"%d", &arr[i].marks[j]);
					sum += arr[i].marks[j];
				}
				arr[i].avg = (float)sum/marks_number;

			} 
			return (size_t)arr;
		} else return 1;
	} else return 0;

}

void writeFile(student *arr, unsigned n)
{
	size_t i = 0, j;
	FILE *file;
	
	file = fopen("output.txt", "w");
	
	for (; i < n; i++) 
	{
		fprintf(file, "%s", arr[i].name);
		for (j = 0; j < n; j++)
			fprintf(file, "%d ", arr[i].marks[j]);
		fprintf(file, "%.2f\n", arr[i].avg);
	}	

	fclose(file);
}

int main()
{
	student *arr = (student *) malloc (sizeof(student));
	unsigned n;
	size_t p;
	char type;

	switch (p = readFile(arr, &n))
	{
		case 0: printf("%s", "File not opened."); break;
		case 1: printf("%s", "Not enought memory."); break;
		default:
			arr = (student *)p;
			printf("%s\n", "[L]exicographical or [N]onincreasing?");
			scanf("%c", &type);
			switch (type)
			{
				case 'N':
					sort(arr, n, sizeof(student), &nonincreasing);
				break;
				case 'L':
					sort(arr, n, sizeof(student), &lexicographical);
				break;
			}
			writeFile(arr, n);
		break;
	}	

	getchar();

	return 0;
}